sympy/polys/rings.py

```diff
@@ -3005,3 +3179,302 @@ def sqf_list(f, all=False):
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms(polynomials, ring, domain)
+                return [gcd], None
+
+            syms = pi.free_variables()
+            if allsame and syms != common:
+                allsame = False
+            common &= syms
+
+        # The loop is complete if they all have the same symbols.
+        if allsame:
+            return polynomials, common
+
+        # Extract coefficients as polys containing only the common symbols.
+        all_polys = []
+        for i, pi in enumerate(polynomials):
+            coeffs_i = pi.coeff_split(pi.free_variables() - common)
+            all_polys.extend(coeffs_i)
+
+            # Quick exit:
+            if any(len(c) == 1 for c in coeffs_i):
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms((all_polys + polynomials[i+1:]), ring, domain)
+                return [gcd], None
+
+def gcd_prs(polynomials):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x - y, x - y
+    >>> gcd_prs(polynomials)
+    x - y
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in polynomials):
+        return gcd_terms(polynomials, ring, domain)
+
+    polynomials, monom_gcd = monomial_extract(polynomials)
+
+    polynomials, common_symbols = _gcd_preprocess_polys(polynomials)
+
+    gcd = polynomials[0]
+    for pi in polynomials[1:]:
+        gcd = _gcd_prs(gcd, pi)
+        if gcd == domain.one:
+            break
+
+    if monom_gcd is not None:
+        gcd = gcd * monom_gcd
+
+    return gcd
+
+def _gcd_prs(p1, p2):
+    """
+    Returns the greatest common divisor (GCD) of two polynomials using the
+    Polynomial Resultant Sequences (PRS) method.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import _gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = 4*x**2 + 8*x + 10
+    >>> g = 2*x**2 + 6*x + 10
+    >>> _gcd_prs(f, g)
+    2
+
+    """
+    x = p1.main_variable()
+
+    c1, pp1 = cont_prim(p1, x)
+    c2, pp2 = cont_prim(p2, x)
+
+    h = pp1.subresultants(pp2, x)[-1]
+    c = gcd_prs([c1, c2])
+
+    ring = p1.ring
+    domain = ring.to_domain()
+
+    _, h = cont_prim(h, x)
+
+    if domain.canonical_unit(h.coeff_wrt(x, h.degree(x))):
+        h = -h
+
+    h = h * c
+
+    return h
+
+def _gcd_(f, g):
+    """Helper function for _gcd_I method."""
+
+    old_ring = f.ring
+    old_domain = old_ring.domain
+
+    if not old_domain.is_Exact:
+        try:
+            new_domain = old_domain.get_exact()
+        except DomainError:
+            return old_domain.one, f, g
+
+        new_domain = old_domain.get_exact()
```

Comment:
We don't need to call `get_exact` twice. Can you give an example when `DomainError` is raised?